### PR TITLE
Raw input improvements

### LINKED ISSUE #1333: Window doesn't focus at startup on macOS
https://github.com/opentk/opentk/blob/develop/src/OpenTK/Platform/MacOS/CocoaNativeWindow.cs#L268 This call in OpenTK's CocoaNativeWindow instantiation seems to be failing since the window is not getting focused (Window.Focused == false).

When this happens, the following call to `setActive` doesn't result in the expected state:

screen shot 2018-01-17 at 5 04 50 pm

### PR DESCRIPTION
**Pull request by peppy**
- Fixes raw input not working correctly on macOS.
- Fixed absolute raw input devices (tablets etc.) not working correctly.

Closes #1333 via OpenTK update.

### CODE REVIEW

**Comment by FreezyLemon** [comment]
Testing on a Win10, 64 Bit PC.
Tablet: Wacom CTL-480S, driver version 6.3.9w5.

With raw input on, the mouse works fine.
The tablet has very specific behavior:
When the game window is active and the cursor is inside the window, the game tries to map the absolute position given by the tablet to itself assuming that it spans the whole screen (even across monitors: When I disconnect a monitor, the 'area' inside the window changes).
It is a bit hard to explain, I hope the video will clear it up a bit more.

What is interesting is that the "Screen Area" in the driver does not change this, though the "Tablet Area" does influence it.

A small bug I found is that when
- entering or exiting the window on any border or
- changing device (moving the mouse after the tablet pen was active or vice versa)

the game seems to activate mouse scrolls (up and down) randomly.

Videos:
[Mouse](https://streamable.com/srkok)
[Tablet](https://streamable.com/gh9an)

**Comment by LittleEndu** [comment]
Scrolling inside settings and beatmap carousel doesn't work when trying to click and drag with tablet.

**Comment by lukis101** [comment]
Win10 1709, H420 using "PenTablet 12.3.7.170420"  
Same behavior observed: Entire virtual desktop space is mapped to game screen.
Area and left-hand option seem work correctly.  
With "Support TabletPC" checked cursor gets mapped to a small top-left area (with some offset affected by window position)

**Comment by peppy** [comment]
Lazer doesn't have "support tabletpc". Are you testing on... stable? Please don't report here in that case.

**Comment by peppy** [comment]
@LittleEndu removing your feedback as you provided no details on where/how you're testing

**Comment by LittleEndu** [comment]
I was testing on lazer client after checking out this branch. I think the relevant class is called ScrollContainer. I noticed it in the settings menu and in the beatmap carousel. 
You can click and drag on its background and it will scroll, kinda like you would do on a phone if you scroll something. This works fine with mouse but nothing happened with tablet (the ScrollContainer stood still)

**Comment by peppy** [comment]
@FreezyLemon @LittleEndu please test again with recent additions. as far as i can tell it should be 1:1 with stable behaviour.

File: osu.Framework/Input/Handlers/Mouse/OpenTKRawMouseHandler.cs [hunk 1]
```diff
@@ -56,109 +54,110 @@ public override bool Initialize(GameHost host)
...
-                            {
-                                newStates.Add(s);
-                                mostSeenStates = i;
-                            }
+                            var newStates = new List<OpenTK.Input.MouseState?>();
 
-                            i++;
-                        }
+                            int i = 0;
+                            while (i <= mostSeenStates + 1)
```
Comment (Aergwyn) [inline osu.Framework/Input/Handlers/Mouse/OpenTKRawMouseHandler.cs]: Why isn't this `for(int i = 0; i <= mostSeenStates + 1; i++)`?

File: osu.Framework/Input/Handlers/Mouse/OpenTKRawMouseHandler.cs [hunk 2]
```diff
@@ -56,109 +54,110 @@ public override bool Initialize(GameHost host)
...
+                                Vector2 currentPosition;
 
-                            if (state.IsAbsolute)
-                            {
-                                const int raw_input_resolution = 65536;
+                                if (lastState != null && state.Equals(lastState.RawState))
+                                    continue;
 
-                                if (mapAbsoluteInputToWindow)
+                                if (state.IsAbsolute)
```
Comment (Aergwyn) [inline osu.Framework/Input/Handlers/Mouse/OpenTKRawMouseHandler.cs]: How about moving this block into a method that returns the calculated position? This makes this part less complex and also moves the declaration of `currentPosition` after the continue which feels nicer.

**Comment by LittleEndu** [comment]
I found out the issue that I was having with ScrollContainer. It seems raw input doesn't mix inputs from different devices correctly. I use physical touchpad click (not tap to click) instead of pen touch (because I have my nib removed) and now I saw that the cursor jumps to seemingly random position for just a frame and that's why noting works. It seems to be only touchpad click tho as mouse click with pen positioning works fine. So in the end, it's not a problem with ScrollContainer, it's problem with everything that needs to be clicked.

**Comment by LittleEndu** [comment]
Everything else worked just fine with my Huion 420 (with Support TabletPC disabled). Tell me if more testing is needed.

**Comment by peppy** [comment]
Definitely on the latest commit? Because that should fix what you are mentioning

**Comment by LittleEndu** [comment]
I'm on [f4a1c2bf...](https://github.com/peppy/osu-framework/commit/f4a1c2bf3b144e60d62dc5179258fbed233d5715) now but the issue with touchpad click not clicking where tablet's cursor was is still present.

**Comment by peppy** [comment]
muilti-monitor systems should now behave correctly.

File: osu.Framework/Input/Handlers/Mouse/OpenTKRawMouseHandler.cs [hunk 3]
```diff
@@ -52,80 +59,137 @@ public override bool Initialize(GameHost host)
...
+                            while (lastStates.Count < newStates.Count)
+                                lastStates.Add(null);
 
-                                // update the windows cursor to match our raw cursor position.
-                                // this is important when sensitivity is decreased below 1.0, where we need to ensure the cursor stays withing the window.
-                                var screenPoint = host.Window.PointToScreen(new Point((int)currentPosition.X, (int)currentPosition.Y));
-                                OpenTK.Input.Mouse.SetPosition(screenPoint.X, screenPoint.Y);
+                            for (int i = 0; i < newStates.Count; i++)
+                            {
+                                if (newStates[i]?.IsConnected != true)
```
Comment (smoogipoo) [inline osu.Framework/Input/Handlers/Mouse/OpenTKRawMouseHandler.cs]: Can OpenTK even give us null values here? I thought GetState would return an empty state?

File: osu.Framework/Input/MouseState.cs [hunk 4]
```diff
@@ -46,7 +46,7 @@ public IMouseState Clone()
         {
             var clone = (MouseState)MemberwiseClone();
             clone.buttons = new List<MouseButton>(buttons);
-            clone.LastState = null;
+            clone.LastState = LastState;
```
Comment (smoogipoo) [inline osu.Framework/Input/MouseState.cs]: Have you made sure that this isn't causing any sort of leaks due to stored references?

Comment (peppy) [reply osu.Framework/Input/MouseState.cs]: WIll double-check, but the main dereferencing is still present in `InputManager`

File: osu.Framework/Input/Handlers/Mouse/OpenTKRawMouseHandler.cs (hunk 3 above)
Comment (peppy) [reply osu.Framework/Input/Handlers/Mouse/OpenTKRawMouseHandler.cs]: I don't believe so...

**Comment by lukis101** [comment]
Finally dug into device-side HID descriptors and programmed a microcontroller for testing this. Will be able to emulate different input types, but for now only did Pen/Digitizer class (used by tablets).
Tested on:  game [f9cba37](https://github.com/ppy/osu/commit/f9cba37a6564c029971375faf4df1ee81e0d87c3), framework [ee4021c](https://github.com/ppy/osu-framework/pull/1371/commits/ee4021ce25a4e0af968ca69f58884a03f4a00245), Win 10, single monitor (harder to show with multiple)

With H420 and Huion driver app running, all seems good...
Without the driver app - when input comes from a "Pen" HID type, windows shows "dot" cursor and input is mapped to top-left of game screen as demonstrated in video.
Microcontroller moves mouse in fullscreen zigzags (code and HID descriptor (barely) shown on the right). In last third of the video, movement and clicks are from my H420.
[Video](https://streamable.com/yfqww)

P.S. sorry for the confusion before ._.

**Comment by peppy** [comment]
Yeah, in that mode we need to handle windows ink, which has not been added yet.

File: osu.Framework/Input/InputManager.cs [hunk 5]
```diff
@@ -258,11 +258,8 @@ protected virtual void HandleNewState(InputState state)
             // state potentially from a different input source.
             if (last.Mouse != null && state.Mouse != null)
             {
-                if (last.Mouse.GetType() == state.Mouse.GetType())
-                {
-                    last.Mouse.LastState = null;
+                if (state.Mouse.LastState == null && last.Mouse.GetType() == state.Mouse.GetType())
```
Comment (smoogipoo) [inline osu.Framework/Input/InputManager.cs]: Can we have a comment as to why this null check is here?

Comment (peppy) [reply osu.Framework/Input/InputManager.cs]: yep

**Pull request merged**